[Alternate WebM Player] Switch from internal to experimental feature

youssefsoli opened the pull request:
#### c369a87a8d7f4bf40edc7627d1d8b9e7e2dff938
<pre>
[Alternate WebM Player] Switch from internal to experimental feature
<a href="https://bugs.webkit.org/show_bug.cgi?id=242165">https://bugs.webkit.org/show_bug.cgi?id=242165</a>
&lt;rdar://96191749&gt;

Reviewed by Eric Carlson.

Switching the alternate WebM player from internal to experimental in order
to eexpose the feature to Safari Technology Preview.

The feature is explicitly disabled on test runners in order to continue
having test coverage for the old WebM player.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetPreferencesToConsistentValues):

Canonical link: <a href="https://commits.webkit.org/253995@main">https://commits.webkit.org/253995@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09fa6419691035e8bfff6bf3ff5304fcfc3dc614

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87643 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31730 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18400 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96849 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150611 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91614 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30088 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26215 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79742 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91601 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93261 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-15-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74402 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24134 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79468 "Found 1 new API test failure: TestWebKitAPI.VideoControlsManager.VideoControlsManagerMultipleVideosScrollPlayingVideoWithSoundOutOfView") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67150 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79459 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27787 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13294 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73162 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27750 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14310 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26032 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2810 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29443 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37197 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75993 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29349 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33587 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16850 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode") | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -387,13 +405,123 @@ ResourceLoaderOptions loaderOptions(
+    if (!newPixelBuffer)
+        return false;
+
+    m_lastPixelBuffer = newPixelBuffer;
```
WTFMove

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -387,13 +405,123 @@ ResourceLoaderOptions loaderOptions(
+
 void MediaPlayerPrivateWebM::setNaturalSize(FloatSize size)
 {
     FloatSize oldSize = m_naturalSize;
```
auto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -387,13 +405,123 @@ ResourceLoaderOptions loaderOptions(
     FloatSize oldSize = m_naturalSize;
     m_naturalSize = size;
     if (oldSize != m_naturalSize) {
         INFO_LOG(LOGIDENTIFIER, "was ", oldSize.width(), " x ", oldSize.height(), ", is ", size.width(), " x ", size.height());
```
`FloatSize` has a `LogArgument` template, so you should be able to write this as 

`        INFO_LOG(LOGIDENTIFIER, "was ", oldSize, ", is ", size);`

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -472,14 +618,48 @@ ResourceLoaderOptions loaderOptions(
+    if (shouldEnsureLayer()) {
+        destroyDecompressionSession();
+        ensureLayer();
+    } else {
```
An early return instead of an else would be more idiomatic.

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -578,6 +770,9 @@ ResourceLoaderOptions loaderOptions(
     }
+    
+    if (m_readyState < MediaPlayer::ReadyState::HaveCurrentData)
+        setReadyState(MediaPlayer::ReadyState::HaveEnoughData);
```
Shouldn't this be `HaveCurrentData`?

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -734,14 +939,22 @@ ResourceLoaderOptions loaderOptions(
+        if (m_decompressionSession) {
+            m_decompressionSession->requestMediaDataWhenReady([this, trackId] {
+                didBecomeReadyForMoreSamples(trackId);
+            });
```
Ditto

eric-carlson commented on Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
```diff
@@ -602,6 +797,11 @@ ResourceLoaderOptions loaderOptions(
+            m_decompressionSession->requestMediaDataWhenReady([this, trackId] {
+                didBecomeReadyForMoreSamples(trackId);
+            });
+        }
```
Can we guarantee that the player will still be valid when the lambda is called? If not, you should probably use a WeakPtr like you do for the callbacks from the audio renderer and the AVSBDL

eric-carlson reviewed (approved):
r=me with some nits

eric-carlson reviewed (approved):
r=me

youssefsoli committed:
[Alternate WebM Player] Switch from internal to experimental feature
https://bugs.webkit.org/show_bug.cgi?id=242165
<rdar://96191749>

Reviewed by Eric Carlson.

Switching the alternate WebM player from internal to experimental in order
to eexpose the feature to Safari Technology Preview.

The feature is explicitly disabled on test runners in order to continue
having test coverage for the old WebM player.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetPreferencesToConsistentValues):

Canonical link: https://commits.webkit.org/253995@main
